EIPS/eip-BN256_HashToCurve.md

```diff
@@ -0,0 +1,342 @@
+    end
+end function
+```
+
+In HashToG2, we first map to the underlying twist curve
+and then clear the cofactor.
+Here is the pseudocode for HashToG2:
+
+```
+function HashToG2(m)
+    t00 = HashToBase(m, 0x04, 0x05)
+    t01 = HashToBase(m, 0x06, 0x07)
+    t10 = HashToBase(m, 0x08, 0x09)
+    t11 = HashToBase(m, 0x0a, 0x0b)
+    t0 = (t00, t01)
+    t1 = (t10, t11)
+    h0 = BaseToTwist(t0)
+    h1 = BaseToTwist(t1)
+    h = ECAdd(h0, h1)
+    g = ECMul(h, r)     # Clear cofactor; g is now an element of G2
+    return g
+end function
+```
+
+Here is the pseudocode for BaseToTwist; the only real difference
+between BaseToG1 and BaseToTwist are the different constants
+and the different finite fields.
+All of these operations are performed in the finite field
+of the twist curve.
+As in BaseToG1, `inverse`, `is_square`, and `sqrt` perform
+the same operations in the underlying finite field GF(p^2).
+Our algorithms for `is_square` and `sqrt` come from this
+[paper](https://eprint.iacr.org/2012/685).
+
+```
+function BaseToTwist(t)
+    # All operations are done in the finite field GF(p^2)
+    # Here, the twist curve satisfies the equation
+    #       y^2 == g'(x) == x^3 + b'
+    p1 = (−1 + sqrt(−3))/2
+    p2 = −3
+    p3 = 1/3
+    p4 = g'(1)
+    s = (p4 + t^2)^3
+    alpha = inverse(t^2*(p4 + t^2))
+    x1 = p1 − p2*t^4*alpha
+    x2 = 1 − x1
+    x3 = −1 − p3*s*alpha
+    a1 = x1^3 + b'
+    a2 = x2^3 + b'
+    r1 = is_square(a1)
+    r2 = is_square(a2)
+    i = (r1 − 1)*(r2 − 3)/4 + 1
+    c1 = CTEq(1, i)
+    c2 = CTEq(2, i)
+    c3 = CTEq(3, i)
+    x = c1*x1 + c2*x2 + c3*x3
+    y = sign0(t)*sqrt(x^3 + b')
+    return (x, y)
+end function
+```
+
+## Rationale
+The current design choices reflect one choice of hashing algorithm.
+There is freedom in choosing the HashToBase function
+and this could easily be changed.
+Within HashToBase, the particular hashing algorithm
+(Keccak256 in our case) could also be modified.
+It may be desired to change the call to `sign0`
+at the end with `is_square`, as this would result
+in the same deterministic map to curve from the
+Fouque and Tibouchi
+[paper](https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf)
+and ensure HashToG1 is indifferentiable from a random oracle;
+they proved this result in their paper.
+It may be possible to show that switching the `is_square`
+call with `sign0` does not affect indifferentiability,
+although this has not been proven at this time.
+
+[EIP-2537](./eip-2537.md)
+would allow for similar BLS signatures to be computed on
+the BLS12-381 elliptic curve.
+This proposal calls for additional precompiled contracts
+to allow for full use of pairing-based cryptography with
+this elliptic curve.
+This is the same rationale for wanting to add a hash-to-curve
+algorithm for BN256.
+
+## Backwards Compatibility
+There are no backward compatibility concerns.
+
+## Test Cases
+No underlying algorithms are being changed and so there is no need
+to specify additional test cases.
```

Comment:
Since this is a cross client consensus change, it feels like there would be significant value to including test cases (inputs paired with expected output) so that clients can validate their implementation likely aligns with other clients prior to running integration tests.

This is not necessary for a draft EIP, but strongly advised for consensus EIPs at some point in the process.

Reply:
I am surprised this would be "considered a cross client consensus change". I believe the hash-to-curve algorithms would not touch consensus at all, as I am only looking to add precompile contracts to the EVM. Am I mistaken or misunderstanding something?

If specific test cases would be included (such as known results for BaseToG1 or BaseToG2), those could be included without much hassle. Is this what you are suggesting?

Reply:
New precompiles are a consensus change in that if clients don't agree on the output of a precompile given ea particular input then they will desync and we'll no longer have consensus across clients.

Reply:
I will add some test cases in the near future. Yes, having consistent outputs across clients is critical and I now understand why tests are needed.